### ISSUE TITLE
Parallelism Variables

### ISSUE THREAD

**Issue by yitzchak** [ninja-build/ninja#2125]
Having access to the number of threads that an edge has access to via a variable would be useful for rules that are multithreaded. For example, if `config.parallelism = 16` and the pool has a depth of 2 then each edge can use 8 threads safely. One could do something like this then:

```
rule foobar
  command = quux -j $jobs $in $out
```

I can submit a PR to this effect, just curious if this would even be entertained.

**Comment by jonesmz** [comment]
Some pull requests get merged in days, others languish for years.

If it's easy to make a PR, just do it, and people will discuss your code changes.

**Comment by yitzchak** [comment]
That's not particularly helpful comment.

I did not say it was easy to make a PR, I asked if the concept would be entertained since ninja's authors seems to desire a minimalist approach.

**Comment by jonesmz** [comment]
I wasn't trying to be unhelpful. It's just the reality of the project. You won't get an answer to this question that goes deeper than "don't know, lets see what it looks like".

**Comment by yitzchak** [comment]
Okay, thanks.

**Comment by hadrielk** [comment]
You may also want to look through previous issues about this topic, like #2029.

I'm also not clear on what your proposal means. Are you saying when invoking `quux`, ninja would fill in the value of `$jobs` for that argument? Would the value of that be a dynamic number, or a fixed number that literally is just the number of overall `-j` jobs divided by that rule's pool's depth?

If the former (dynamic), I don't think that will work for you, because an instant after it spawns that job, it can spawn N more jobs.

If the latter (fixed), then as an FYI: it's something that you can already do if you're using CMake to generate your ninja rules. (although it does require a cmake re-config to change the values, not a command-line switch to `ninja`)

**Comment by yitzchak** [comment]
@hadrielk Thanks for the reply. We are not using CMake but our own metabuilder. We are basically looking for dynamic control of the thread count of edges that contain parallel processes.

Several of the edges in our system are parallel compilers which are in a console pool. They don't know what Ninja's value for `-j` is so we have to configure the job count for those parallel compilers in the metabuild stage, before ninja is actually run. What we would like to do is be able to pass the current value of `-j` to the parallel compiler edges so they know how many threads to use.

None of these parallel compilers have a pool depth > 1, but this seems a dynamic value of `$jobs` should respect the pool depth.  In this case `$jobs` would be equal to the value of `-j` divided by the pool depth.

**Comment by hadrielk** [comment]
OK, so it looks like what you want is a "fixed" number - fixed in the sense that the value doesn't change during the course of building, but it does potentially change each time `ninja` is invoked because it depends on the `-j` jobs value it's using.

You've probably already considered this, but just in case not...

Since you're writing your own metabuilder, have you considered just writing the value into a rules file each time, but not the "main" rules file... instead, separating those particular aspects into a separate rules file and `include`ing it in your main rules file - to make writing it trivial and very fast? (by `include` I don't mean subninja, I mean the `include` statement in ninja rules, which I don't think are documented but are widely used in practice)

In other words declare your own variables in a separate file - one you can overwrite with your metabiuilder or even just a simple front-end bash or python or whatever script. And resolve those variables in your main rules file, by having your metabuilder create the rules originally with them as variables (i.e., `$num_jobs` or whatever).

Obviously that means you'd have to use such a script to invoke ninja instead of invoking it directly, but if you're looking for a solution in some reasonable timeframe...

**Comment by eli-schwartz** [comment]
It sounds like you might be more interested in #1139? You want `quux` to be able to fork as many jobs as it is allowed to based on ninja -j values, but I'd assume you also want ninja to know how many jobs `quux` is using, to prevent double-dipping.

The solution is a cooperative scheduler such as the GNU make jobserver. Even client support is sufficient to run GNU make as the top-level scheduler and let `ninja` and `quux` both request job slots from make.
